src/transactions/feed/SwapFeedItem.test.tsx

```diff
@@ -0,0 +1,125 @@
+import { render } from '@testing-library/react-native'
+import React from 'react'
+import { Provider } from 'react-redux'
+import { ReactTestInstance } from 'react-test-renderer'
+import { RootState } from 'src/redux/reducers'
+import SwapFeedItem from 'src/transactions/feed/SwapFeedItem'
+import {
+  Fee,
+  TokenAmount,
+  TokenExchangeMetadata,
+  TokenTransactionTypeV2,
+} from 'src/transactions/types'
+import { createMockStore, getElementText, RecursivePartial } from 'test/utils'
+import { mockCeurAddress, mockCusdAddress } from 'test/values'
+
+const MOCK_TX_HASH = '0x006b866d20452a24d1d90c7514422188cc7c5d873e2f1ed661ec3f810ad5331c'
+
+describe('SwapFeedItem', () => {
+  function renderScreen({
+    storeOverrides = {},
+    inAmount,
+    outAmount,
+    metadata = {},
+    fees = [],
+  }: {
+    inAmount: TokenAmount
+    outAmount: TokenAmount
+    metadata?: TokenExchangeMetadata
+    fees?: Fee[]
+    storeOverrides?: RecursivePartial<RootState>
+  }) {
+    const store = createMockStore({
+      ...storeOverrides,
+    })
+
+    const tree = render(
+      <Provider store={store}>
+        <SwapFeedItem
+          exchange={{
+            __typename: 'TokenExchangeV2',
+            type: TokenTransactionTypeV2.SwapTransaction,
+            transactionHash: MOCK_TX_HASH,
+            timestamp: 1234,
+            block: '2345',
+            inAmount,
+            outAmount,
+            metadata,
+            fees,
+          }}
+        />
+      </Provider>
+    )
+
+    return {
+      store,
+      ...tree,
+    }
+  }
+
+  function expectDisplay({
+    getByTestId,
+    expectedTitleSections,
+    expectedSubtitleSections,
+    expectedIncomingAmount,
+    expectedOutgoingAmount,
+  }: {
+    getByTestId: (testId: string) => ReactTestInstance
+    expectedTitleSections: string
+    expectedSubtitleSections: string
+    expectedIncomingAmount: string
+    expectedOutgoingAmount: string
+  }) {
+    const title = getElementText(getByTestId('SwapFeedItem/title'))
+    expect(getElementText(title)).toEqual(expectedTitleSections)
+
+    const subtitle = getElementText(getByTestId('SwapFeedItem/subtitle'))
+    expect(getElementText(subtitle)).toEqual(expectedSubtitleSections)
+
+    const amountDisplay = getByTestId('SwapFeedItem/incomingAmount')
+    expect(getElementText(amountDisplay)).toEqual(expectedIncomingAmount)
+
+    const tokenDisplay = getByTestId('SwapFeedItem/outgoingAmount')
+    expect(getElementText(tokenDisplay)).toEqual(expectedOutgoingAmount)
+  }
+
+  it('renders correctly for cUSD to cEUR swap', async () => {
+    const { getByTestId } = renderScreen({
+      inAmount: {
+        tokenAddress: mockCeurAddress,
+        value: 2.93,
+      },
+      outAmount: {
+        tokenAddress: mockCusdAddress,
+        value: 2.87,
+      },
+    })
+    expectDisplay({
```

Comment:
this abstraction seems a little odd to me, i think for tests it is okay to be a little verbose rather than try to abstract away logic.

as an aside, i would recommend simplifying this test to read something like "expect swapScreen.title to exist, expect [text] to exist, ...." and avoid looking for test id's. i understand that you are trying to test that the correct text appears in the correct "box" marked by the test id, but i think that still doesn't always guarantee that the display is correct anyway (e.g. if the styles are messed up, then even though the test-id's and texts match, they could still be positioned in the wrong order relative to other things / not visible to the user) so in that sense, i think it's not too important to test for.

i like this article https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-querying-by-text - maybe this helps. Kent (the author of this article) is one of the authors of the react testing library. this article is more geared towards web than mobile, but i think the concept of testing for what the user sees rather than technical implementation detail is still relevant :)

Reply:
Thanks for sharing the article. It was fun to read!

I removed the `expectDisplay()` function. I used it because the other feed item test codes used it.

I tried to do "expect swapScreen.title to exist", but I was not able to find a way to do it.